src/hotspot/share/ci/ciEnv.cpp

```diff
@@ -1265,7 +1268,7 @@ void ciEnv::record_method_not_compilable(const char* reason, bool all_tiers) {
     _compilable = new_compilable;
 
     // Reset failure reason; this one is more important.
-    _failure_reason = nullptr;
+    _failure_reason.clear();
     record_failure(reason);
```

Comment:
@dean-long the tricky part is here, I think:
We are in `ciEnv::record_method_not_compilable`. Depending on above condition, we either update the failure reason, or don't. So there may already be an old failure reason (the first one) that we overwrite (with a second one) or leave.

This method is called from `Compile::record_method_not_compilable`. There, we also call `Compile::record_failure`, which sets `Compile::_failure_reason`, with no such condition.

Hence, there can be multiple failure reasons around, I think. And I cannot fit multiple messages in a single buffer. Thus, I need some way to dynamically allocate.

Now: is this good or desirable? I don't know. Potentially having different failure messages at Compile and ciEnv - not sure if that is reasonable. We should probably discuss that. But for now I'm just trying to fix the code the way it works.

What do you think?

Reply:
I think initially we care only about "most" important failure because implementation allowed to report only one failure.
If you can keep all failures AND report all of them it would be nice, I think. May be in separate RFE.

We should have only one place where we keep failures.

Reply:
Reporting all failures might make sense for logging, but we can only report one reason, unless we want to do something like chained exceptions, which I don't think we need.  The different layers is already looking like a custom exception handling mechanism, unfortunately.

Reply:
Piling on... I opt for keeping this code simple. A list of subsequent failure states could easily be part of logging.

Reply:
Okay, I am voting for UL to print all failures we hit during compilation and keep reporting only one reason at the end.
```
log_error(compilation) (failiure_reason);
```
Or `log_warning()`

Reply:
@vnkozlov @dean-long @tstuefe 
Ok, I mean logging of additional failure reasons might even be done in a separate RFE, right?
But don't we already log the failures? I mean there is the `log()->elem(` code.

And should I still keep the "most important" failure reason?